text/0171-cloudfront-redesign.md

```diff
@@ -0,0 +1,484 @@
+
+## Lambda@Edge
+
+Lambda@Edge is an extension of AWS Lambda, a compute service that lets you execute functions that customize the content that CloudFront delivers. You
+can author Node.js or Python functions in the US East (N. Virginia) region, and then execute them in AWS locations globally that are closer to the
+viewer, without provisioning or managing servers. Lambda@Edge functions are associated with a specific behavior and event type. Lambda@Edge can be
+used rewrite URLs, alter responses based on headers or cookies, or authorize requests based on headers or authorization tokens.
+
+By default, Lambda@Edge functions are attached to the default behavior:
+
+```ts
+const myFunc = new lambda.Function(...);
+const myDist = new cloudfront.Distribution(...);
+myDist.addLambdaFunctionAssociation({
+  functionVersion: myFunc.currentVersion,
+  eventType: EventType.VIEWER_REQUEST,
+});
+```
+
+Lambda@Edge functions can also be associated with additional behaviors, either at behavior creation or after the fact, either by attaching
+directly to the behavior, or to the distribution and referencing the behavior.
+
+```ts
+// Assigning at behavior creation.
+myOrigin.addBehavior('/images/*.jpg', {
+  viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+  defaultTtl: cdk.Duration.days(7),
+  functionAssociation: [{
+    functionVersion: myFunc.currentVersion,
+    eventType: EventType.VIEWER_REQUEST,
+  }]
+});
+
+// Assigning after creation.
+const myImagesBehavior = myOrigin.addBehavior('/images/*.jpg', ...);
+myImagesBehavior.addFunctionAssociation({
+  functionVersion: myFunc.currentVersion,
+  eventType: EventType.VIEWER_REQUEST,
+});
+
+myDist.addFunctionAssociation({
+  functionVersion: myFunc.currentVersion,
+  eventType: EventType.ORIGIN_REQUEST,
+  behavior: myImagesBehavior,
+});
+```
+
+---
+
+# Motivation
+
+The existing aws-cloudfront module doesn't adhere to standard naming convention, lacks convenience methods for more easily interacting with
+distributions, origins, and behaviors, and has been in an "experimental" state for years. This proposal aims to bring a friendlier, more ergonic
+interface to the module, and advance the module to a GA-ready state.
+
+# Design Summary
+
+The approach will create a new top-level Construct (`Distribution`) to replace the existing `CloudFrontWebDistribution`, as well as new constructs
+to represent the other logical resources for a distribution (i.e., `Origin`, `Behavior`). The new L2s will be created in the same aws-cloudfront
+module and no changes will be made to the existing L2s to preserve the existing experience. Unlike the existing L2, the new L2s will feature a
+variety of convenience methods (e.g., `addBehavior`) to aid in the creation of the distribution, and provide several out-of-the-box defaults for
+building distributions off of other resources (e.g., buckets, load balanced services).
```

Comment:
I would mention the fact that this new construct is somewhat optimized for the common use-case of creating a distribution with a single origin, reducing the cognitive load on the user.